### ISSUE TITLE
Segfault during zts_join

### ISSUE THREAD

**Issue by YFL** [zerotier/libzt#64]
Hi everyone,
I'm on a Linux machine (manjaro, Linux 5.4.35-1) and I'm using the debug build of libzt from `libzt/lib/debug/linux-x86_64/`. I've got an issue with libzt throwing a `SEGFAULT` during the execution of `zts_join`. I've enabled every LWIP debug layer by changing every `LWIP_DBG_OFF` to `LWIP_DBG_ON` in `lwipopts.h` inside `libzt/src`. The output is as follows:
```
sys_timeout: 0x7f4114000bd0 abs_time=79291506 handler=lwip_cyclic_timer arg=0x7f411b1225f8
sys_timeout: 0x7f4114000c00 abs_time=79291506 handler=lwip_cyclic_timer arg=0x7f411b122610
sys_timeout: 0x7f4114000c30 abs_time=79291506 handler=lwip_cyclic_timer arg=0x7f411b122628
sys_timeout: 0x7f4114000c60 abs_time=79291506 handler=lwip_cyclic_timer arg=0x7f411b122640
sys_timeout: 0x7f4114000c90 abs_time=79290606 handler=lwip_cyclic_timer arg=0x7f411b122658
sct calling h=lwip_cyclic_timer t=0 arg=0x7f411b122658
tcpip: mld6_tmr()
sys_timeout: 0x7f4114000c90 abs_time=79290706 handler=mld6_tmr arg=0x7f411b122658
sct calling h=mld6_tmr t=1 arg=0x7f411b122658
tcpip: mld6_tmr()
sys_timeout: 0x7f4114000c90 abs_time=79290806 handler=mld6_tmr arg=0x7f411b122658
sct calling h=mld6_tmr t=0 arg=0x7f411b122658
tcpip: mld6_tmr()
sys_timeout: 0x7f4114000c90 abs_time=79290906 handler=mld6_tmr arg=0x7f411b122658
sct calling h=mld6_tmr t=0 arg=0x7f411b122658
tcpip: mld6_tmr()
sys_timeout: 0x7f4114000c90 abs_time=79291006 handler=mld6_tmr arg=0x7f411b122658
sct calling h=mld6_tmr t=0 arg=0x7f411b122658
tcpip: mld6_tmr()
sys_timeout: 0x7f4114000c90 abs_time=79291106 handler=mld6_tmr arg=0x7f411b122658
sct calling h=mld6_tmr t=0 arg=0x7f411b122658
tcpip: mld6_tmr()
sys_timeout: 0x7f4114000c90 abs_time=79291206 handler=mld6_tmr arg=0x7f411b122658
sct calling h=mld6_tmr t=1 arg=0x7f411b122658
tcpip: mld6_tmr()
sys_timeout: 0x7f4114000c90 abs_time=79291306 handler=mld6_tmr arg=0x7f411b122658
netif_ip6_addr_set: netif address being changed
netif: IPv6 address 1 of interface  set to FD12:AC4A:1E71:C0BB:6C99:9394:3865:FC89/0x00
netif_ip6_addr_set_state: netif address state being changed
netif: IPv6 address 0 of interface  set to FE80::200:FF:FE00:0/0x08
netif: IPv6 address 0 of interface  set to FE80::200:FF:FE00:0/0x08
netif_ip6_addr_set_state: netif address state being changed
netif: IPv6 address 1 of interface  set to FD12:AC4A:1E71:C0BB:6C99:9394:3865:FC89/0x08
./RUN: line 3: 51785 Segmentation fault      (core dumped) ./test ../app/config
```
This happens on a ZeroTier network with IPv4 AutoAssign disabled and IPv6 RFC 4193 enabled, which is public. The above output is generated with the identity and other files already present. The initial execution yields a much longer output, where the following repeats for a loooooooong time:
```
sct calling h=lwip_cyclic_timer t=0 arg=0x7f0aa9c3f658
tcpip: mld6_tmr()
sys_timeout: 0x7f0aa0000c90 abs_time=79719160 handler=mld6_tmr arg=0x7f0aa9c3f658
sct calling h=mld6_tmr t=0 arg=0x7f0aa9c3f658
tcpip: mld6_tmr()
sys_timeout: 0x7f0aa0000c90 abs_time=79719260 handler=mld6_tmr arg=0x7f0aa9c3f658
sct calling h=mld6_tmr t=0 arg=0x7f0aa9c3f658
tcpip: mld6_tmr()
sys_timeout: 0x7f0aa0000c90 abs_time=79719360 handler=mld6_tmr arg=0x7f0aa9c3f658
sct calling h=mld6_tmr t=0 arg=0x7f0aa9c3f658
tcpip: mld6_tmr()
sys_timeout: 0x7f0aa0000c90 abs_time=79719460 handler=mld6_tmr arg=0x7f0aa9c3f658
sct calling h=mld6_tmr t=0 arg=0x7f0aa9c3f658
tcpip: mld6_tmr()
sys_timeout: 0x7f0aa0000c90 abs_time=79719560 handler=mld6_tmr arg=0x7f0aa9c3f658
sct calling h=mld6_tmr t=0 arg=0x7f0aa9c3f658
tcpip: mld6_tmr()
sys_timeout: 0x7f0aa0000c90 abs_time=79719660 handler=mld6_tmr arg=0x7f0aa9c3f658
sct calling h=mld6_tmr t=0 arg=0x7f0aa9c3f658
tcpip: mld6_tmr()
sys_timeout: 0x7f0aa0000c90 abs_time=79719760 handler=mld6_tmr arg=0x7f0aa9c3f658
sct calling h=mld6_tmr t=0 arg=0x7f0aa9c3f658
tcpip: mld6_tmr()
sys_timeout: 0x7f0aa0000c90 abs_time=79719860 handler=mld6_tmr arg=0x7f0aa9c3f658
sct calling h=mld6_tmr t=0 arg=0x7f0aa9c3f658
tcpip: mld6_tmr()
sys_timeout: 0x7f0aa0000c90 abs_time=79719960 handler=mld6_tmr arg=0x7f0aa9c3f658
sct calling h=lwip_cyclic_timer t=0 arg=0x7f0aa9c3f5f8
tcpip: ip_reass_tmr()
sys_timeout: 0x7f0aa0000bd0 abs_time=79720960 handler=ip_reass_tmr arg=0x7f0aa9c3f5f8
sct calling h=lwip_cyclic_timer t=0 arg=0x7f0aa9c3f610
tcpip: etharp_tmr()
etharp_timer
sys_timeout: 0x7f0aa0000c00 abs_time=79720960 handler=etharp_tmr arg=0x7f0aa9c3f610
```
The SEGFAULT occurs in the `lwip_prepare_netif_status_msg` function. The stacks bottom is `ZeroTier::_zts_run_service`. Currently the value of the tap variable in `lwip_prepare_netif_status_msg` is 0x0, but at another time I was inspecting the situation, that value was non-null.

The part of my program being executed is as follows:
```c++
#include <iostream>
#include <fstream>

#include <unistd.h>
#include <signal.h>

#include <arpa/inet.h>
#include <ZeroTier.h>

#include <zt_lua_wrap.h>
#include <config_reader.h>

static int node_ready = false, network_ready = false;

auto callback_func(zts_callback_msg *msg)
{
    if(msg->eventCode == ZTS_EVENT_NODE_ONLINE)
    {
        std::cout << "Node " << std::hex << msg->node->address << " is online" << std::endl;
        node_ready = true;
        return;
    }
    if(msg->eventCode == ZTS_EVENT_NODE_UNRECOVERABLE_ERROR)
    {
        std::cout << "Node_unrecoverable_error occured" << std::endl;
        return;
    }
    if(msg->eventCode == ZTS_EVENT_NETWORK_OK)
    {
        std::cout << "Network " << std::hex << msg->network->nwid << " OK" << std::endl;
        return;
    }
    if(msg->eventCode == ZTS_EVENT_NETWORK_READY_IP4)
    {
        std::cout << "Network ready for IP4. We don't care, we wait for IPv6" << std::endl;
        //network_ready = true;
        return;
    }
    if(msg->eventCode == ZTS_EVENT_NETWORK_READY_IP6)
    {
        std::cout << "Network ready for IP6" << std::endl;
        network_ready = true;
        return;
    }
    if(msg->eventCode == ZTS_EVENT_NETWORK_REQUESTING_CONFIG)
    {
        std::cout << "Requesting network configuration" << std::endl;
        return;
    }
    if(msg->eventCode == ZTS_EVENT_PEER_P2P)
    {
        std::cout << std::hex << msg->peer->address << " ready for P2P" << std::endl;
        return;
    }
}

auto ctrl_c_handler(int signal) -> void
{
    zts_stop();
    exit(1);
}

auto main(int argc, char *argv[]) -> int
{
    if(argc != 2)
    {
        std::cerr << "Please specify the path to the config file as the only argument" << std::endl;
        return 1;
    }

    struct sigaction sig;
    sig.sa_handler = ctrl_c_handler;
    sigaction(SIGINT, &sig, nullptr);

    const char *conf_path = argv[1];
    std::ifstream conf(conf_path);

    if(conf.good())
    {
        std::map<std::string, std::string> conf_map = zt_lua::read_conf(conf);
        std::cout << "Starting" << std::endl;
        zts_start("zt_runtime", callback_func, ZTS_DEFAULT_PORT);
        while(!node_ready) { sleep(1); }

        std::cout << "Joining" << std::endl;
        std::cout << "network_id in conf_map: " << conf_map["network_id"] << std::endl;
        uint64_t nwid = std::stoull(conf_map["network_id"], 0, 16);
        std::cout << "network_id: " << std::hex << nwid << std::endl;
        zts_join(nwid);
        while(!network_ready) { sleep(1); }
        

...
```

If I can supply more information about the issue, just tell me what and how can I find it.

Thank you in advance!

**Comment by YFL** [comment]
More information that came to my mind:
The device is shown in the ZeroTier Central, it's even shown as online for some time. (I'm sure I've seen "via RELAY" in the status message under the nodeId in the beta UI for managing connected nodes, not sure about "via P2P")

**Comment by YFL** [comment]
Another one:
using sudo zerotier-cli join <nwid> succeeds, and the node is shown in the ZeroTier Central

**Comment by joseph-henry** [comment]
Thanks for bringing this to my attention. I believe this is fixed in the most recent tagged version `1.3.3` https://github.com/zerotier/libzt/commit/a0b50530d37d9c13d30a68bf1d4686485be36327

In this latest version there are a couple of changes to be aware of:
 - Main C API header has been renamed to `ZeroTierSockets.h`
 - Error reporting scheme and values have changed slightly (see header file)
 - Event code values have changed significantly (see header file)

Let me know if this issue persists or you discover anything else.
